# Data Consistency Checks

In [10]:
# Import libraries 
import pandas as pd
import numpy as np
import os

In [11]:
# Set path
path = r'C:\Users\Ryan Pares\Instacart Basket Analysis'

In [12]:
# Import dataframes ('products.csv' and 'orders_wrangled.csv')
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

# 1. Consistency Checks

In [13]:
# Data consistency check by reviewing descriptive statistics
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number     order_dow  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01  2.776219e+00   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01  2.046829e+00   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00  0.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00  1.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01  3.000000e+00   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01  5.000000e+00   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02  6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

In [14]:
df_ords.drop(columns=['Unnamed: 0'], inplace =True)

In [15]:
df_ords.describe()

order_id       user_id  order_number     order_dow  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01  2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01  2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00  0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00  1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01  3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01  5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02  6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

# 2. Mixe-Type Data

In [16]:
# Create a dataframe
df_test = pd.DataFrame()

In [17]:
# Create a mixed type column
df_test['mix'] = ['a', 'b', 1, True]

In [18]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [19]:
# Check for any mixed-type columns
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


#### Mixed data types are found, Single data type for column is decided (as String)

In [20]:
# Change data type
df_test['mix'] = df_test['mix'].astype('str')

In [21]:
df_test['mix']

0       a
1       b
2       1
3    True
Name: mix, dtype: object

# 3. Missing Values

In [22]:
#Functions used:'isnull()', which is used the find missing observations & the 'sum()' function returns the total missing values
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [23]:
# To view the 16 missing values, create a subset of the dataframe containing only the values in question

In [24]:
# Create a new dataframe, df_nan, containing on the values within the 'product_name' column that meets the condition of 'isnull()' true
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [25]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

# Addressing missing values

In [26]:
# There are several ways to deal with missing data:
# 1. Create a new variable that acts like a flag based on the missing value
# 2. Impute the value with the mean or median of the column (if the variable is numeric)
# 3. Remove or filter out the missing data

In [27]:
# Looking at df_nan it's clear imputation isn't an option because the data-type is a string
# You can either remove the missing values entirely or filter the non-missing values into a subset dataframe

In [28]:
# Compare the number of rows in the current dataframe with the number in your subset once the missing rows are removed
df_prods.shape

(49693, 5)

In [29]:
# Create a new dataframe, setting the 'isnull()' condition to False so that only non-missing values are present
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [30]:
# Check the number of rows again to make sure it descreased
df_prods_clean.shape

(49677, 5)

In [31]:
# To drop only NaNs from a particular column use the following command
df_prods.dropna(subset = ['product_name'], inplace = True)

In [32]:
df_prods.isnull().sum()

product_id       0
product_name     0
aisle_id         0
department_id    0
prices           0
dtype: int64

# 5. Duplicates

In [33]:
# Look for full duplicates within the dataframe
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [34]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

# 6. Addressing Duplicates

In [35]:
#Remove duplicates using 'df.drop_duplicates()' 
#Row count of df_prods_clean before duplicate removal
df_prods_clean.shape

(49677, 5)

In [36]:
# Create a new dataframe that does not include duplicates 
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [37]:
#Row count of df_prods_clean after duplicate removal
df_prods_clean_no_dups.shape

(49672, 5)

# 7. Tidying up & exporting changes

In [38]:
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'products_checked.csv'))

# Task

# 2. Run df.describe()

In [39]:
df_ords.describe()

order_id       user_id  order_number     order_dow  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01  2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01  2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00  0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00  1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01  3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01  5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02  6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

In [40]:
#Examine 'order_number', 'orders_day_of_week', 'days_since_prior_order' for data insights.
#'Order_number' peaks at 100, indicating a high volume within a certain period.
#'Orders_day_of_week' ranges from 0 to 6, hinting at a zero-indexed week starting Monday or Sunday.
#Treat 'order_id', 'user_id', 'order_number' as strings, not numbers, for data integrity.
#Zeros in 'days_since_prior_order' imply same-day orders or initial customer orders.
#Lower counts in 'days_since_prior_order' may point to data omissions.
#A 30-day cap on 'days_since_prior_order' suggests a monthly ordering cycle.

# 3. Check for Mixed-data type

In [41]:
# Check for any mixed-type columns
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
         print (col)

In [42]:
df_ords.dtypes

order_id                    int64
user_id                     int64
order_number                int64
order_dow                   int64
order_hour_of_day           int64
days_since_prior_order    float64
dtype: object

# 4. No mixed data type present in df_ords

# 5. Check for missing values

In [43]:
df_ords.isnull().sum()

order_id                       0
user_id                        0
order_number                   0
order_dow                      0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

In [44]:
# 'days_since_prior_order' column is missing 206209 values
# Due to the large number of missing values, we should preserve it instead of removing

# 6. Address missing values

In [45]:
# Create a subset of the data from the original dataframe
df_ords_nan = df_ords[df_ords['days_since_prior_order'].isnull() == True]

In [46]:
df_ords_nan

order_id  user_id  order_number  order_dow  order_hour_of_day  \
0         2539329        1             1          2                  8   
11        2168274        2             1          2                 11   
26        1374495        3             1          1                 14   
39        3343014        4             1          6                 11   
45        2717275        5             1          3                 12   
...           ...      ...           ...        ...                ...   
3420930    969311   206205             1          4                 12   
3420934   3189322   206206             1          3                 18   
3421002   2166133   206207             1          6                 19   
3421019   2227043   206208             1          1                 15   
3421069   3154581   206209             1          3                 11   

         days_since_prior_order  
0                           NaN  
11                          NaN  
26                          NaN  
39                          NaN  
45                          NaN  
...                         ...  
3420930                     NaN  
3420934                     NaN  
3421002                     NaN  
3421019                     NaN  
3421069                     NaN  

[206209 rows x 6 columns]

In [47]:
# Check whether all 'order_number' values equal to 1 correspond to missing values in 'days_since_prior_order' column
# Filter the dataframe where 'order_number' equals 1
df_first_orders = df_ords_nan[df_ords_nan['order_number'] == '1']

In [48]:
# Check values in 'days_since_prior_order'
df_first_orders_missing = df_first_orders['days_since_prior_order'].isna().all()

In [49]:
# Print result
print("All first orders have missing 'days_since_prior_order':", df_first_orders_missing) 

All first orders have missing 'days_since_prior_order': True


In [50]:
#Creating column to flag order numbers
df_ords['first_order'] = df_ords['days_since_prior_order'].isna()

In [51]:
df_ords

order_id  user_id  order_number  order_dow  order_hour_of_day  \
0         2539329        1             1          2                  8   
1         2398795        1             2          3                  7   
2          473747        1             3          3                 12   
3         2254736        1             4          4                  7   
4          431534        1             5          4                 15   
...           ...      ...           ...        ...                ...   
3421078   2266710   206209            10          5                 18   
3421079   1854736   206209            11          4                 10   
3421080    626363   206209            12          1                 12   
3421081   2977660   206209            13          1                 12   
3421082    272231   206209            14          6                 14   

         days_since_prior_order  first_order  
0                           NaN         True  
1                          15.0        False  
2                          21.0        False  
3                          29.0        False  
4                          28.0        False  
...                         ...          ...  
3421078                    29.0        False  
3421079                    30.0        False  
3421080                    18.0        False  
3421081                     7.0        False  
3421082                    30.0        False  

[3421083 rows x 7 columns]

In [52]:
# Check data-type of new column and update if necessary
df_ords['first_order'].dtype

dtype('bool')

In [53]:
# Re-check data to make sure totals are accurate
df_ords.isnull().sum()

order_id                       0
user_id                        0
order_number                   0
order_dow                      0
order_hour_of_day              0
days_since_prior_order    206209
first_order                    0
dtype: int64

##### Missing values have been addressed, by creating a column that flags order numbers as either True/False. Removing/filtering these values would not be correct as the missing data is relevant information. 

# 7. Check for duplicate values in df_ords data

In [54]:
df_ords_dup = df_ords[df_ords.duplicated()]

In [55]:
df_ords_dup

Empty DataFrame
Columns: [order_id, user_id, order_number, order_dow, order_hour_of_day, days_since_prior_order, first_order]
Index: []

#### There are no Duplicates in the df_ords data frame.

# 8. Address the duplicates

In [56]:
df_prods_clean_no_dups.describe()

product_id      aisle_id  department_id        prices
count  49672.000000  49672.000000   49672.000000  49672.000000
mean   24850.349775     67.762442      11.728942      9.993282
std    14340.705287     38.315784       5.850779    453.615536
min        1.000000      1.000000       1.000000      1.000000
25%    12432.750000     35.000000       7.000000      4.100000
50%    24850.500000     69.000000      13.000000      7.100000
75%    37268.250000    100.000000      17.000000     11.100000
max    49688.000000    134.000000      21.000000  99999.000000

#### No duplicates.

# Revision 8. Address the duplicates

In [58]:
df_prods_clean_no_dups.describe()

product_id      aisle_id  department_id        prices
count  49672.000000  49672.000000   49672.000000  49672.000000
mean   24850.349775     67.762442      11.728942      9.993282
std    14340.705287     38.315784       5.850779    453.615536
min        1.000000      1.000000       1.000000      1.000000
25%    12432.750000     35.000000       7.000000      4.100000
50%    24850.500000     69.000000      13.000000      7.100000
75%    37268.250000    100.000000      17.000000     11.100000
max    49688.000000    134.000000      21.000000  99999.000000

##### The maximum value in the 'prices' column is unusually high '99999'. It's likely such an extreme outlier is either a placeholder or and error
##### The 'product_id' column has a maximum value of 49688. If 49672 is intended to represent an equal number of distinct products( from shape), this may indicate the presence of an error or the possibility of repeated entries.

In [59]:
df_ords.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_cleaned.csv'))

In [60]:
# Number of unique product ids
df_unique_product_ids = df_prods_clean_no_dups['product_id'].nunique()

In [62]:
print(f"The number of unique product IDs: {df_unique_product_ids}")

The number of unique product IDs: 49670


In [64]:
# Check if product_id values are sequential & without gaps
df_all_ids = set(range(1, df_prods_clean_no_dups['product_id'].max() + 1))

In [65]:
# Create a set of actual ids in the dataframe
df_actual_ids = set(df_prods_clean_no_dups['product_id'].unique())

In [66]:
# Find any missing ids
missing_ids = df_all_ids - df_actual_ids

In [67]:
if missing_ids:
    print(f"There are missing product IDs: {sorted(missing_ids)}")
else:
    print("No missing product IDs; they are sequential.")

There are missing product IDs: [34, 69, 116, 262, 525, 1511, 1780, 2240, 2586, 3159, 3230, 3736, 4283, 4790, 6799, 26519, 38183, 40440]


In [68]:
# The maximum value of 99999 also needs to be investigated
outliers = df_prods_clean_no_dups[df_prods_clean_no_dups['prices'] == 99999]

In [69]:
print(outliers)

       product_id           product_name  aisle_id  department_id   prices
33666       33664  2 % Reduced Fat  Milk        84             16  99999.0


In [72]:
# Removing the error since price is skewed.
df_prods_extra_clean_no_dups = df_prods_clean_no_dups[df_prods_clean_no_dups['prices'] != 99999]

In [73]:
df_prods_extra_clean_no_dups.describe()

product_id      aisle_id  department_id        prices
count  49671.000000  49671.000000   49671.000000  49671.000000
mean   24850.172334     67.762115      11.728856      7.980256
std    14340.795118     38.316100       5.850806     66.952504
min        1.000000      1.000000       1.000000      1.000000
25%    12432.500000     35.000000       7.000000      4.100000
50%    24850.000000     69.000000      13.000000      7.100000
75%    37268.500000    100.000000      17.000000     11.100000
max    49688.000000    134.000000      21.000000  14900.000000

In [74]:
# Investigating the rows with the highest prices to see the other outliers
df_prods_sorted = df_prods_extra_clean_no_dups.sort_values(by='prices', ascending=False)

In [75]:
top_priced_products = df_prods_sorted.head()

In [76]:
print(top_priced_products)

       product_id                      product_name  aisle_id  department_id  \
21554       21553  Lowfat 2% Milkfat Cottage Cheese       108             16   
9020         9020  Boneless Skinless Chicken Thighs        35             12   
40490       40486                   Chicken Tenders        49             12   
21468       21467            Wild Caught Raw Shrimp        15             12   
25580       25579     Naturally Smoked Trout Fillet        15             12   

        prices  
21554  14900.0  
9020      25.0  
40490     25.0  
21468     25.0  
25580     25.0  


#### Product priced 14900  must be removed as the outlier.

In [77]:
# Remove Outlier
df_prods_extra_clean_no_dups_2 = df_prods_extra_clean_no_dups[df_prods_extra_clean_no_dups['prices'] != 14900]

In [78]:
df_prods_extra_clean_no_dups_2.describe()

product_id      aisle_id  department_id        prices
count  49670.000000  49670.000000   49670.000000  49670.000000
mean   24850.238716     67.761305      11.728770      7.680437
std    14340.931850     38.316060       5.850834      4.199381
min        1.000000      1.000000       1.000000      1.000000
25%    12432.250000     35.000000       7.000000      4.100000
50%    24850.500000     69.000000      13.000000      7.100000
75%    37268.750000    100.000000      17.000000     11.100000
max    49688.000000    134.000000      21.000000     25.000000

#### Anomalies removed and descriptive statistics appears normal.

# 9. Export 

In [81]:
# Export as Products re-checked & Orders_cleaned.csv
df_ords_dup.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'products_re-checked.csv'))
df_ords.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_cleaned.csv'))